### PR TITLE
Validation section

### PR DESCRIPTION
**Pull request by ofivite**
This is a PR which adds a section on validation methods. Structurally, it is organised into 3 parts: before, throughout and after training. Each of the parts contains subsections covering corresponding topics, ranging from data preprocessing to overfitting and performance validation. The PR represents a further developed version of the initial proposal shown in [this presentation](https://indico.cern.ch/event/1033476/contributions/4340052/attachments/2245654/3808275/validation_update_17May.pdf) and is ready to be reviewed.

### CODE REVIEW

File: content/validation/intro.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+# Introduction
+
+In general, ML models don't really work out of the box: for example, mostly often it is not sufficient to simply instantiate the model class, call its `fit()` method followed by `predict()` and proceed straight to the inference step of the analysis:
```
Comment (aperloff) [inline content/validation/intro.md:3]: "...box. For example, most often it is not sufficient to simply instantiate the model class, call its `fit()` method followed by `predict()`, and then proceed straight to the inference step of the analysis."

File: content/validation/intro.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+X_train, X_test, y_train, y_test = \
+        train_test_split(X, y, test_size=.4, random_state=42)
+
+clf = SVC(kernel="linear", C=0.025)
+clf.fit(X_train, y_train)
+print(f'Accuracy: {clf.score(X_test, y_test)}')
+# Accuracy: 0.4
+```
+
+Being an extremely simplified and naive example, one would be lucky to have the code above producing valid and optimal model. This is because it explicitly doesn't check for those things which could've gone wrong and therefore is prone to producing undesirable results. Indeed, there are several pitfalls which one may encounter on the way towards implementation of ML into their analysis pipeline. These can be easily avoided by being aware of those and performing a few simple checks here and there.
```
Comment (aperloff) [inline content/validation/intro.md:20]: "producing" -> "produce a "

File: content/validation/intro.md [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+Therefore this section is intended to **overview potential issues on the ML side and how they can be approached** in order to train a robust and optimal model. The section is designed to be to a large extent analysis-agnostic, focusing on general and most common validation steps from ML perspective, without paying particular emphasis on the physical context. However, for illustrative purposes it will be supplemented with some examples from HEP and additional links for further reading.  As the last remark, in the following there will mostly an emphasis on the validation items specific to _supervised learning_. This includes classification and regression problems as being so far the most common use cases amongst HEP analysts.
```
Comment (aperloff) [inline content/validation/intro.md:22]: Add a comma after "Therefore"

Comment (aperloff) [reply content/validation/intro.md:22]: "overview" -> "review"

Comment (aperloff) [reply content/validation/intro.md:22]: "...to be, to a large extent, analysis-agnostic. It will focus on common, generalized validation steps..."

Comment (aperloff) [reply content/validation/intro.md:22]: Add a comma after "illustrative purposes"

File: content/validation/before/domains.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
+Data plays a crucial role in the process of training any ML model. It is something from which the model learns to solve a given task and therefore care needs to be taken with its handling:
+
+1. By making sure that a data set is relevant to the problem and no deviation from expectations is introduced: so that a resulted model doesn't incorporate these undesired discrepancies.
```
Comment (aperloff) [inline content/validation/before/domains.md:3]: "introduced. This makes sure that the model doesn't incorporate undesired fluctuations or features."

File: content/validation/before/domains.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+2. By performing a proper preprocessing of the data set: so that the training step goes smoothly.
```
Comment (aperloff) [inline content/validation/before/domains.md:4]: Remove the ":"

File: content/validation/before/domains.md [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+In this section a general [domain perspective](https://en.wikipedia.org/wiki/Domain_adaptation) on data will be covered and in the following sections a more granular look will be taken from the side of features and construction of inputs to the model.
```
Comment (aperloff) [inline content/validation/before/domains.md:6]: "will be covered. In the following sections..."

File: content/validation/before/domains.md [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+## Coverage  
+To begin with, one needs to bear in mind that training data should be as close as possible to data they expect to have in the context of analysis. Speaking in more formal terms,
+> Domains of training (used to train the model) and inference (used to make final predictions) data sets should not sizeably diverge.
+
+??? example "Examples"
```
Comment (aperloff) [inline content/validation/before/domains.md:12]: Why did you put "??? example"? Do you need help with something or is this just a mistake?

File: content/validation/before/domains.md [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    * Consider the model which is trained to predict the energy of a hadron given its energy deposits in the calorimeter (represented e.g. in the form of image or graph). Data consists of the showers initiated by a particle generated by a particle gun and having discrete values of energies (e.g. 1 GeV, 10 GeV, 20 GeV, etc.). However, in the real world settings, the model will be applied to showers produced by particles with underlying continuous energy spectrum. Although ML models are known for their capability to interpolate beyond their training domain, without apropriate tests model performance in the parts of the energy spectrum outside of its training domain is not a priori clear.  
+
+**How to deal with**
```
Comment (aperloff) [inline content/validation/before/domains.md:16]: '### Solution'

**Review by aperloff — COMMENTED** [review]
This is just my first set of review comments. There will me more to come.

File: content/validation/before/domains.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
+Data plays a crucial role in the process of training any ML model. It is something from which the model learns to solve a given task and therefore care needs to be taken with its handling:
+
+1. By making sure that a data set is relevant to the problem and no deviation from expectations is introduced. This makes sure that the model doesn't incorporate undesired fluctuations or features.
+1. By performing a proper preprocessing of the data set so that the training step goes smoothly.
```
Comment (savvy379) [inline content/validation/before/domains.md:4]: This should be 2 instead of 1. In general I think you could frame this a little more, maybe in the intro sentence say something like 'there are two main considerations when collecting and preparing data for an ML task

File: content/validation/before/domains.md [hunk 10]
```diff
@@ -0,0 +1,39 @@
+Data plays a crucial role in the process of training any ML model. It is something from which the model learns to solve a given task and therefore care needs to be taken with its handling:
+
+1. By making sure that a data set is relevant to the problem and no deviation from expectations is introduced. This makes sure that the model doesn't incorporate undesired fluctuations or features.
```
Comment (savvy379) [inline content/validation/before/domains.md:3]: not clear what expectations means here, could you be more specific?

File: content/validation/before/domains.md [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+### Solution    
+
+It is a widely known practice to introduce **scale factor** (SF) **corrections** to account for possible discrepancies between data and MC simulation. Effectively, that means that the model is probed on some part of the domain on which it wasn't trained on (data) and then corrected for any differences by using a meaningful set of observables to derive SFs. One particularly promising approaches to remedy for data/MC domain difference is to use adversarial approaches to fully leverage the multidimensionality of the problem, as described in a [DeepSF note](https://cds.cern.ch/record/2666647).
+
+Another solution would be to incorporate methods of [**domain adaptation**](https://github.com/zhaoxin94/awesome-domain-adaptation) into an ML pipeline, which essentially guide the model to be invariant and robust towards domain shift. Particularly in HEP, a [Learning to Pivot with Adversarial Networks](https://arxiv.org/pdf/1611.01046.pdf) paper was one of the pioneers to investigate how a pile-up dependency can be mitigated, which can also be easily expanded to building a model robust to domain shift[^1].
```
Comment (savvy379) [inline content/validation/before/domains.md:20]: I think this point is a little over simplified...I would maybe not say it's easy to expand to building a model robust to domain shift, but maybe say this is an example of how specific dependencies can be mitigated

File: content/validation/before/domains.md (hunk 9 above)
Comment (savvy379) [inline content/validation/before/domains.md:4]: This should be point 2. In general I think you can frame these points a little better as well, maybe in the intro sentence say something like 'there are two main concerns when collecting and preparing data'

File: content/validation/before/domains.md (hunk 10 above)
Comment (savvy379) [inline content/validation/before/domains.md:3]: not clear what expectations means here, could you be more specific?

File: content/validation/before/domains.md [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+Another solution would be to incorporate methods of [**domain adaptation**](https://github.com/zhaoxin94/awesome-domain-adaptation) into an ML pipeline, which essentially guide the model to be invariant and robust towards domain shift. Particularly in HEP, a [Learning to Pivot with Adversarial Networks](https://arxiv.org/pdf/1611.01046.pdf) paper was one of the pioneers to investigate how a pile-up dependency can be mitigated, which can also be easily expanded to building a model robust to domain shift[^1].  
+
+Last but not the least, a usage of [**bayesian neural networks**]() has a great advantage of getting uncertainties estimate along with each prediction. If these uncertainties are significantly larger for some samples, this could indicate that they come from the domain beyond the training one (a so-called [out-of-distribution samples](https://ai.googleblog.com/2019/12/improving-out-of-distribution-detection.html)) - and therefore care needs to be taken in that cases.
```
Comment (savvy379) [inline content/validation/before/domains.md:22]: can you link to the documentation section on bayesian neural networks here? and maybe provide a link to a paper or example that describes what kind of care would need to be taken?

File: content/validation/before/domains.md [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+
+## Population
+Furthermore, nowadays analyses are searching for very rare processes and therefore are interested in low-populated regions of the phase space. And even though the domain of interest can be covered in the training data set, it may also happen that not sufficiently in terms of data points populating those regions. That makes the model behaviour on an event which falls into those regions unpredictable - because it couldn't learn how to generalise in those areas due to a lack of data to learn from. Therefore,
```
Comment (savvy379) [inline content/validation/before/domains.md:25]: this sentence 'it may also happen that not sufficiently' needs to be rewritten, it's missing something like 'may also not be be sufficiently covered' or something, I'm not sure exactly what you're trying to say here.

File: content/validation/before/domains.md (hunk 9 above)
Comment (yaourtpourtoi) [reply content/validation/before/domains.md:4]: This is actually a feature, one can just write "1." all the time in numbered lists and it will render correctly as 1., 2., 3. etc.

**Comment by yaourtpourtoi** [comment]
Thank you @savvy379 for the comments! I incorporated all the suggestions and clarifications.

**Comment by aperloff** [comment]
I am giving this PR a conditional approval. I think that it's 90% of the way there. There needs to be a few changes with respect to links, LaTeX, and some emphasis within the text. However, we think that those changes are best accomplished in a second PR.
